Reduce noise from NotEntitledException logging

prdoyle opened the pull request:
The stack trace tends to attract attention even when entitlements are not the cause of a problem, so let's avoid the stack trace, but still include stack frame info from the frame of interest. We're in something of a unique position here, where we know the exact frame that matters (except in cases where the problem stems from landing on the wrong frame, I suppose).

Let's use child loggers so we can filter these independently.

prdoyle committed:
Refactor: findRequestingFrame

prdoyle committed:
INFO instead of WARN for NotEntitledException.

Some of these are expected, so an INFO seems more appropriate.

The stack trace tends to attract attention even when entitlements are not the
cause of a problem, so let's avoid the stack trace, but still include stack
frame info from the frame of interest.

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

rjernst commented:
> Some of these are expected
Instead of making the unexpected ones look "ok" with info level, can we hide the expected ones? eg can we create a child logger based on the module name + entitlement (or some other identifier) that we can mute with normal logging filtering? Then GCS/AWS libs can ignore those known warnings without flooding the ES log, but expected entitlement failures can be properly warned.

rjernst commented:
> The stack trace tends to attract attention even when entitlements are not the cause of a problem
If we filter out the "known" entitlements failures we want to ignore, is that still the case? We wanted to ensure the log contains everything we need to fix the issue, even if the exception is caught and ignored.

prdoyle commented:
> Instead of making the unexpected ones look "ok" with info level, can we hide the expected ones?

That's not what's happening here. We can't know a priori which ones are expected, so we've decided to log  them all at `info` and rely on another mechanism to report a more serious failure for unexpected ones.

Also, teaching the server which plugins have expected entitlement failures seems backward; the server doesn't know about the plugins. I could imagine adding a way to put this into the policy files though.

prdoyle committed:
Merge branch 'main' into chill-bro

prdoyle commented:
Ok I'm not sure I see the advantage of that approach. Emitting warnings when everything is operating as intended seems bad. Perhaps I'll understand better after our sync meeting today.

prdoyle committed:
Merge branch 'main' into chill-bro

prdoyle commented:
Thinking about it some more, I like the "child logger" idea irrespective of `info` vs `warn`. The only wrinkle is that our alert currently matches on the exact logger name, so I'll need to figure out how to generalize that, but it should be doable.

prdoyle commented:
Note to self: if the module has no name, we could use the same `ALL_UNNAMED` tag we use elsewhere.

prdoyle committed:
Use child loggers for Not Entitled logs

rjernst commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyManager.java:
```diff
@@ -538,11 +538,13 @@ private void notEntitled(String message, Class<?> callerClass) {
         // Don't emit a log for muted classes, e.g. classes containing self tests
         if (mutedClasses.contains(callerClass) == false) {
+            var moduleName = callerClass.getModule().getName();
+            var notEntitledLogger = LogManager.getLogger(PolicyManager.class + "." + ((moduleName == null) ? "ALL_UNNAMED" : moduleName));
```
nit: we use `ALL-UNNAMED` (with a dash) in the policy files

ldematte replied:
Why the module name and not the component name? (plugin name, "server" or "apm-agent"?)
Wouldn't we have clash for ALL-UNNAMED from different plugins if we need to adjust log level for one non-modular plugin?

rjernst replied:
The component name is also reasonable, but the finer grained we can make this the less changing this logging could affect other log messages.

prdoyle replied:
whoops

ldematte replied:
Having both as agreed is the best 👍  Thanks!

prdoyle committed:
Use warn, and include compoenent name

prdoyle committed:
Fix ALL_UNNAMED

prdoyle committed:
Mute entitlement warnings from repositories

rjernst commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyManager.java:
```diff
@@ -530,15 +535,23 @@ public void checkWriteProperty(Class<?> callerClass, String property) {
+                .walk(this::findRequestingFrame)
+                .map(StackFrame::toString)
+                .orElse("");
+            notEntitledLogger.warn(message + frameInfoSuffix);
```
Should there be a newline (with indentation) so it's sort of like a stack trace? What does one of these messages look like?

prdoyle replied:
The issue we were resolving here was that it looked too much like an error, so I figured making it look like an exception stack trace would be a bit counterproductive. Also, we've experienced situations lately where multi-line logs get split up into multiple logs, which is really inconvenient.

But I don't feel too strongly about it.

prdoyle replied:
Here is an example:

```
2025-03-11T12:41:55,308][WARN ][o.e.e.r.p.P.(.o.e.server ] [runTask-0] Not entitled: component [(server)], module [org.elasticsearch.server], class [class org.elasticsearch.bootstrap.Elasticsearch$EntitlementSelfTester], operation [start process]org.elasticsearch.server@9.1.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch$EntitlementSelfTester.ensureCannotStartProcess(Elasticsearch.java:296)
```

In IntelliJ, the file name and line number are a hyperlink.

prdoyle replied:
huh I don't like the lack of spacing there tho

prdoyle replied:
FWIW, with a newline, a tab, and an `at`, it can look like this:

```
[2025-03-11T12:51:03,821][WARN ][o.e.e.r.p.P.(.o.e.server ] [runTask-0] Not entitled: component [(server)], module [org.elasticsearch.server], class [class org.elasticsearch.bootstrap.Elasticsearch$EntitlementSelfTester], operation [start process]
	at org.elasticsearch.server@9.1.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch$EntitlementSelfTester.ensureCannotStartProcess(Elasticsearch.java:296)
```

rjernst commented on modules/repository-s3/src/main/config/log4j2.properties:
```diff
@@ -12,3 +12,7 @@ logger.com_amazonaws_auth_profile_internal_BasicProfileConfigFileLoader.level =
 logger.com_amazonaws_services_s3_internal_UseArnRegionResolver.name = com.amazonaws.services.s3.internal.UseArnRegionResolver
 logger.com_amazonaws_services_s3_internal_UseArnRegionResolver.level = error
+
+logger.org_elasticsearch_entitlement_runtime_policy_PolicyManager.name = org.elasticsearch.entitlement.runtime.policy.PolicyManager.repository-s3
```
Can we only change the specific aws sdk module that is problematic?

prdoyle replied:
It's hard to imagine wanting to get a warning about any module in this component, but I can add the module if you think that's better.

prdoyle commented on modules/repository-gcs/src/main/config/log4j2.properties:
```diff
@@ -0,0 +1,3 @@
+logger.org_elasticsearch_entitlement_runtime_policy_PolicyManager.name = org.elasticsearch.entitlement.runtime.policy.PolicyManager.repository-gcs
```
Ditto.

prdoyle committed:
PR feedback

prdoyle committed:
Merge branch 'main' into chill-bro

prdoyle committed:
Common out the Not Entitled prefix.

We're alerting on this, so let's not rely on every caller of notEntitled to remember it.

rjernst reviewed (approved):
LGTM

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.18](https://github.com/elastic/elasticsearch/pull/124590)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/124590">](https://github.com/elastic/elasticsearch/pull/124590) |
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/124591)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/124591">](https://github.com/elastic/elasticsearch/pull/124591) |
| ✅ |  [9.0](https://github.com/elastic/elasticsearch/pull/124592)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/124592">](https://github.com/elastic/elasticsearch/pull/124592) |
